### PR TITLE
Add ability to preprocess the list of providers

### PR DESCRIPTION
**Pull request by internalsystemerror**
|    Q          |   A
|-------------- | ------
| Documentation | no (not yet)
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | yes
| QA            | yes

### Description

As part of the discussions surrounding https://github.com/laminas/laminas-config-aggregator/issues/25, in order to facilitate this, allowing a list of providers to be "pre-processed" in the same way the merge configuration is "post processed" could be desirable.

As well as for allowing providers to depend on other providers, as in the RFC example, another "pre-processor" could enable the addition/use of development only config providers.

This PR is ready for code review at present, with documentation to be added dependent upon changes from that review.

### CODE REVIEW

**Comment by internalsystemerror** [comment]
Setting this back as a draft as this repo could do with some type improvements prior to adding this (otherwise I'll just end up having to add those as part of this).

Still reviewable for feedback though 👍 .

**Comment by internalsystemerror** [comment]
Added type improvements via https://github.com/laminas/laminas-config-aggregator/pull/29, which this PR is based off.

File: src/ConfigAggregator.php [hunk 1]
```diff
@@ -285,12 +292,29 @@ private function cacheConfig(array $config, null|string $cachedConfigFile): void
...
+     * @param list<PreProcessorCallable|class-string> $processors
+     * @param ProviderIterable $providers
+     * @return ProviderIterable
+     */
+    private function preProcessProviders(array $processors, iterable $providers): iterable
+    {
+        foreach ($processors as $processor) {
+            /** @var PreProcessorCallable $processor */
+            $processor = $this->resolveProcessor($processor);
+            $providers = $processor($providers);
```
Comment (Ocramius) [inline src/ConfigAggregator.php]: IMO to be inlined: `$processor` changes meaning here.

Comment (internalsystemerror) [reply src/ConfigAggregator.php]: Could you clarify what you mean by that? Do you mean from post-processor to pre-processor? Would you just rather a different name for clarity?

Comment (Ocramius) [reply src/ConfigAggregator.php]: I would at least use different variables for the input or the output, to prevent type collisions :)

Comment (internalsystemerror) [reply src/ConfigAggregator.php]: Oh yeah, gotcha 👍, its a copy of the post processing method, I'll update that too

**Review by boesing — APPROVED** [review]
Works for me. 
Please add documentation to this PR as well.

File: src/ConfigAggregator.php [hunk 2]
```diff
@@ -285,14 +292,31 @@ private function cacheConfig(array $config, null|string $cachedConfigFile): void
...
 
+    /**
+     * @param list<PreProcessorCallable|class-string> $processors
+     * @param ProviderIterable $providers
+     * @return ProviderIterable
+     */
+    private function preProcessProviders(array $processors, iterable $providers): iterable
+    {
+        foreach ($processors as $processor) {
+            /** @var PreProcessorCallable $processorCallable */
```
Comment (Ocramius) [inline src/ConfigAggregator.php:303]: Can we somehow move this to the method declaration?

File: src/ConfigAggregator.php [hunk 3]
```diff
@@ -285,14 +292,31 @@ private function cacheConfig(array $config, null|string $cachedConfigFile): void
...
+
     /**
      * @param list<PostProcessorCallable|class-string> $processors
      */
     private function postProcessConfig(array $processors, array $config): array
     {
         foreach ($processors as $processor) {
-            $processor = $this->resolveProcessor($processor);
-            $config    = $processor($config);
+            /** @var PostProcessorCallable $processorCallable */
```
Comment (Ocramius) [inline src/ConfigAggregator.php:317]: Can we somehow move this to the method declaration

**Review by Ocramius — APPROVED** [review]

File: src/ConfigAggregator.php (hunk 2 above)
Comment (internalsystemerror) [reply src/ConfigAggregator.php:303]: iirc, I tried this, but psalm doesn't seem to like a conditional return using custom types

**Comment by internalsystemerror** [comment]
Rebased to use psalm v5, and documentation now added. Ready for review.

**Review by Ocramius — APPROVED** [review]
Excellent, thanks @internalsystemerror!

**Pull request merged**